modules/common/endpoint/endpoint.go

```diff
@@ -70,58 +82,102 @@ func ExposeEndpoints(
 				string(endpointType): "true",
 			},
 		)
-		//
-		// Create the service if none exists
-		//
-		svc := service.NewService(
-			service.GenericService(&service.GenericServiceDetails{
-				Name:      endpointName,
-				Namespace: h.GetBeforeObject().GetNamespace(),
-				Labels:    exportLabels,
-				Selector:  endpointSelector,
-				Port: service.GenericServicePort{
-					Name:     endpointName,
-					Port:     data.Port,
-					Protocol: corev1.ProtocolTCP,
-				}}),
-			exportLabels,
-			timeout,
-		)
-		ctrlResult, err := svc.CreateOrPatch(ctx, h)
-		if err != nil {
-			return endpointMap, ctrlResult, err
-		} else if (ctrlResult != ctrl.Result{}) {
-			return endpointMap, ctrlResult, nil
-		}
-		// create service - end
-
-		// Create the route if none exists
-		// TODO TLS
-		route := route.NewRoute(
-			route.GenericRoute(&route.GenericRouteDetails{
-				Name:           endpointName,
-				Namespace:      h.GetBeforeObject().GetNamespace(),
-				Labels:         exportLabels,
-				ServiceName:    endpointName,
-				TargetPortName: endpointName,
-			}),
-			exportLabels,
-			timeout,
-		)
 
-		ctrlResult, err = route.CreateOrPatch(ctx, h)
-		if err != nil {
-			return endpointMap, ctrlResult, err
-		} else if (ctrlResult != ctrl.Result{}) {
-			return endpointMap, ctrlResult, nil
+		// Create metallb service if specified, otherwise create a route
+		var hostname string
+		if data.MetalLB != nil {
+			annotations := map[string]string{
+				service.MetalLBAddressPoolAnnotation: data.MetalLB.IPAddressPool,
+			}
+			if len(data.MetalLB.LoadBalancerIPs) > 0 {
+				annotations[service.MetalLBLoadBalancerIPs] = strings.Join(data.MetalLB.LoadBalancerIPs, ",")
+			}
+			if data.MetalLB.SharedIP {
+				annotations[service.MetalLBAllowSharedIPAnnotation] = data.MetalLB.IPAddressPool + "-vip"
+			}
+
+			svc := service.NewService(
+				service.MetalLBService(&service.MetalLBServiceDetails{
+					Name:        endpointName,
+					Namespace:   h.GetBeforeObject().GetNamespace(),
+					Annotations: annotations,
+					Labels:      exportLabels,
+					Selector:    endpointSelector,
+					Port: service.GenericServicePort{
+						Name:     endpointName,
+						Port:     data.Port,
+						Protocol: corev1.ProtocolTCP,
+					},
+				}),
+				exportLabels,
+				timeout,
+			)
+			ctrlResult, err := svc.CreateOrPatch(ctx, h)
+			if err != nil {
+				return endpointMap, ctrlResult, err
+			} else if (ctrlResult != ctrl.Result{}) {
+				return endpointMap, ctrlResult, nil
+			}
+			// create service - end
+
+			hostname = svc.GetServiceHostnamePort()
+		} else {
+
+			// Create the service if none exists
```

Comment:
Is that correct? Looks like it will patch an existing service too

Reply:
will update the comment